crates/storage/store.rs

```diff
@@ -15,33 +31,118 @@ use ethrex_common::{
+use ethrex_trie::{Node, NodeRLP};
+use lru::LruCache;
+use rustc_hash::FxBuildHasher;
 use std::{
-    collections::{BTreeMap, HashMap},
-    sync::Mutex,
+    collections::{BTreeMap, HashMap, hash_map::Entry},
+    fmt::Debug,
+    path::{Path, PathBuf},
+    sync::{
+        Arc, Mutex,
+        atomic::AtomicU64,
+        mpsc::{SyncSender, TryRecvError, sync_channel},
+    },
 };
-use std::{fmt::Debug, path::Path};
 use tracing::{debug, error, info};
 /// Number of state trie segments to fetch concurrently during state sync
 pub const STATE_TRIE_SEGMENTS: usize = 2;
 /// Maximum amount of reads from the snapshot in a single transaction to avoid performance hits due to long-living reads
 /// This will always be the amount yielded by snapshot reads unless there are less elements left
 pub const MAX_SNAPSHOT_READS: usize = 100;
 
+// We use one constant for in-memory and another for on-disk backends.
+// This is due to tests requiring state older than 128 blocks.
+// TODO: unify these
+#[allow(unused)]
+const DB_COMMIT_THRESHOLD: usize = 128;
+const IN_MEMORY_COMMIT_THRESHOLD: usize = 10000;
+
+/// Control messages for the FlatKeyValue generator
+#[derive(Debug, PartialEq)]
+enum FKVGeneratorControlMessage {
+    Stop,
+    Continue,
+}
+
+// 64mb
+const CODE_CACHE_MAX_SIZE: u64 = 64 * 1024 * 1024;
+
+// TODO: don't use atomic here, instead wrap in Mutex the whole cache
+#[derive(Debug)]
+struct CodeCache {
+    inner_cache: Mutex<LruCache<H256, Code, FxBuildHasher>>,
+    cache_size: AtomicU64,
+}
+
+impl Default for CodeCache {
+    fn default() -> Self {
+        Self {
+            inner_cache: Mutex::new(LruCache::unbounded_with_hasher(FxBuildHasher)),
+            cache_size: AtomicU64::new(0),
+        }
+    }
+}
+
+impl CodeCache {
+    fn get(&self, code_hash: &H256) -> Result<Option<Code>, StoreError> {
+        let mut cache = self.inner_cache.lock().map_err(|_| StoreError::LockError)?;
+        Ok(cache.get(code_hash).cloned())
+    }
+
+    fn insert(&self, code: &Code) -> Result<(), StoreError> {
+        let mut cache = self.inner_cache.lock().map_err(|_| StoreError::LockError)?;
+        let code_size = code.size();
+        self.cache_size
+            .fetch_add(code_size as u64, std::sync::atomic::Ordering::SeqCst);
+        let cache_len = cache.len() + 1;
+        let mut current_size = self.cache_size.load(std::sync::atomic::Ordering::SeqCst);
+        debug!(
+            "[ACCOUNT CODE CACHE] cache elements (): {cache_len}, total size: {current_size} bytes"
+        );
+
+        while current_size > CODE_CACHE_MAX_SIZE {
+            if let Some((_, code)) = cache.pop_lru() {
+                self.cache_size
+                    .fetch_sub(code.size() as u64, std::sync::atomic::Ordering::SeqCst);
+                current_size = self.cache_size.load(std::sync::atomic::Ordering::SeqCst);
+            } else {
+                break;
+            }
+        }
+
+        cache.get_or_insert(code.hash, || code.clone());
```

Comment:
1. All of this is happening with an exclusive lock taken, which feels like a lot;
2. `fetch_add` already returns the previous value, so it's a bit pointless to read it again just with SeqCst rather than adding again, specially considering with the lock taken nobody else should be modifying the value.